### DISCUSSION TITLE
Bots and their issues - "Treacherous Tides" is bad update for them. The "order system" needs to be reworked.

### DISCUSSION THREAD

**Discussion by JustOneRandomPlayer** [FakeFishGames/Barotrauma#12861] [Ideas and suggestions]
Bots and their issues - "Treacherous Tides" is bad update for them. The "order system" needs to be reworked.

In version 1.0.21.0, bots were more-or-less playable, but after the "Treacherous Tides" update, bots became even worse, and this is due to the order system. I would like to write all this in "Issues", cause bots are one big "issue" and only get worse, but but it does not fit the topic.
I play only in single-player game with bots, so I decided to collect their problems that most hinder the game. Most of it was written on version 1.0.21.0 (I will indicate where exactly), the rest was written after the release of the "Treacherous Tides" update (not many). Maybe some problems have already been fixed, but I couldn’t test the new update for a long time, the bots have become worse, purely because of the orders.

I don't know enough English, so I apologize for the online-translator.

1. - (1.0.21.0, didn't check it in "Treacherous Tides" update) Ice boulders - let the bots to shoot at them. I saw before the explanation that Ice boulders have a lot of HP and they waste ammo, well, that means you need to reduce the HP of Ice boulders, this is not such an important part of the game. It takes longer to "avoid" them than to destroy them (at least in biome, where there are simply too many of them)
While reading the "Treacherous Tides" update I saw that to autopilot added "avoid" ice boulders in the same way as ice spikes, but it’s better to let bots shoot in them - I indicated the reason a little higher.

2. - (1.0.21.0) Monitoring breathing - let the bot-medic pull another character out of the water first and then start performing Cardiopulmonary Resuscitation. Of course, bot-medic does Cardiopulmonary Resuscitation even now, but underwater, - This is not critical, but also not logical.
I'm tested this in "Treacherous Tides" update, - bot-medic saves the characters better. A couple of times he even accidentally pulled the character out of the water, but mostly he does Cardiopulmonary Resuscitation underwater. (it seems that at first he just wanted to go to point for continue “wandering” around the submarine, but he accidentally captured the character. And when he reached the point where he was going (at the same time beating the character against all corners and doors), he began to perform Cardiopulmonary Resuscitation)

3. - (1.0.21.0, didn't check it in "Treacherous Tides" update) Turrets - stop bots from shooting if a character is nearby, at least for Flak Cannon. More than once I was killed by explosions, only because a mob suddenly swam up to me and the AI-bot behind the Flak Cannon shot at it.
3-1. - It also follows, in addition to paragraph 2, that it is worth making sure that if a player or bot dies not far from the submarine, the bot-medic goes to his rescue and carries him into the submarine. This will help in cases where you can’t switch to a bot (multiplayer) or simply don't want to do it (personal challenge).

4. - (1.0.21.0, didn't check it in "Treacherous Tides" update) Complete repair of the submarine in an accessible area - bots just get rid of the leaks and leaving the part of wall damaged (not always, but often). Let bots repair not only the part of the wall where the leak was, but also those that are damaged nearby. Cause now it looks like they’ve repair all, but you walk around and see damaged walls everywhere.

5. - (1.0.21.0 and "Treacherous Tides" update) Bots take too long to aim, even now, after the so-called improvements. Make it dependent on the skill so that at level 100 the bot will shoot the enemy instantly. Players aren't that stupid when they shoot, so why should bots?
Also, bots are too cowardly. Instead of fighting, they run away and wait for someone to come to them - okay, if the enemy AI is human, but the monsters do not chase them through all the doors through which the bots escaped. Moreover, their escape is also useless due to long aiming, because in theory they should retreat and wait for the enemy in order to immediately kill him as soon as he appears, but they wait again (not always, rarely they shoot immediately).

6. - (1.0.21.0, didn't check it in "Treacherous Tides" update) Let the bot-medic treat the Husk Infection, it will at least help in the mission to transport the cultists, I'm already tired of constantly watching them.

7. - (1.0.21.0, didn't check it in "Treacherous Tides" update) Bot priorities - the binding of bots to orders is too strong. A free bot with a welding tool stands two steps away from a leak and, in most cases, does nothing without an order. Or he runs to fix the leak where another bot is already fixing it.

8. - (1.0.21.0, didn't check it in "Treacherous Tides" update) When bots use two-handed weapons, they periodically throw them away. For example, if a bot used an assault rifle in battle and after decided to rid of leaks and immediately switch to a welding tool, then it simply throws away the assault rifle.

9. - (1.0.21.0, didn't check it in "Treacherous Tides" update) When there are enemy captives in handcuffs on board and they are in a room with a reactor, the bot engineer refuses to work with the reactor. It doesn't interfere much, but I wish it didn't happen.

10. - (all versions) Respawnable bots - everything is clear here, make bots respawnable, like players in multiplayer, at least in a single-player game. I'm tired of changing them every time in case of their death, especially considering that they are stupid and can die due to any stupidity. Of course, I can make a mod (someone may have already done it), but it’s better to have it in the game. Moreover, the mod is unlikely to work as expected.

11. - ("Treacherous Tides" update) In version 1.0.21.0, already skilled bots could easily be given the "Assault enemy" order so that they would fight at outposts and everything would be fine. But in the new update they cancel this order when there are no more enemies. Return it as it was in version 1.0.21.0.
11.1. - Why did you completely remove the fact that bots rid of the leaks on enemy objects? In version 1.0.21.0, this helped avoid completely flooding, for example, an enemy outpost when the battle was sloppy and there were too many leaks. Return it as it was in version 1.0.21.0.
11.2. - Now if the bot has an order to use a turret (not only this), then he does not want to go on the attack, but cancels the “Assault enemy” order and returns to the turret. Return it as it was in version 1.0.21.0.
11.3 - Remove this permanent cancellation of orders if bots don't want to carry them out. Let bots then simply not carry them out, but don't cancel it. I'm really tired of constantly repeating all these orders. Again, in version 1.0.21.0 everything was much better and there was no constant cancellation of orders, only rarely canceled.
Here's an example of how orders worked at an enemy outpost in version 1.0.21.0. None of the orders have been cancelled.
![orders before update](https://github.com/Regalis11/Barotrauma/assets/148579030/08c1e738-f24c-4852-b4da-c62759541e16)
And here is an example of how it work after the "Treacherous Tides" update. 
![orders after update](https://github.com/Regalis11/Barotrauma/assets/148579030/b74b4fa3-46b0-4c38-a129-03c6cab34202)
![orders after update_2](https://github.com/Regalis11/Barotrauma/assets/148579030/f841a288-faf7-4d3a-9688-e6e3970e2fa4)
I couldn’t force a one bot to leave the submarine normally at an enemy outpost, but when he left it, he stopped obeying absolutely all orders outside the submarine (he can return to stand, for example, behind the turret, but not attack the enemy).

Maybe it's time to think about reworking the "order system"? The order system in Barotrauma is terrible, it would be much better if there was a "priority system" like in Rimworld, for example. This would help not to have to worry about changing orders for them every time, but to set everything in advance as needed.

But if you are not planning some kind of global update to improve bots, then at least stop making them worse. The "Treacherous Tides" update made the already bad "order system" even worse. In version 1.0.21.0 I could complete the game quite comfortably with bots (despite their issues), but in the "Treacherous Tides" update I don't even want to play with them.

When you make changes to the AI, do you try to play with it yourself? Or are you just testing individual parts of it, without paying attention to how it will end up in the game itself? If the latter, then it's not surprising that when releasing an update with fixes for some parts of AI, others that previously worked fine get worse.
Of course it's good to let players test the game, but not all players will radically test things like AI (at least in multiplayer-oriented games like Barotrauma)
Just in case, let me clarify that this isn't hate or anything else, but just pointing out elements of the AI that are desirable to fix. These are not all bots issues, but those that either most interfere with the comfortable game with them, or are illogical, and which I quickly recalled.

**Comment by bookslayer10** [comment]
Yeah, I don't understand how the bots work with assaulting outposts. I tried to get them to assault it, but once they stepped into the outpost they cancled the command and whenever I tried to re-enable the assault enemies command they just said "No way!"

**Reply by JustOneRandomPlayer** [reply]
Yes, and the most amazing thing is that in version 1.0.21.0 everything worked perfectly. Some bots could be ordered to “Assault Enemy” (it was not canceled as it is now) and they would attack absolutely everyone, while others, which were weaker, could be ordered to “Fight Intruders” and they would attack purely those who were in the submarine.

Even just reading the update, some of the “fixes” raise questions:
_«- Bots don't attempt to repair or fix leaks in non-friendly subs (e.g. wrecks or ruins).»_
They didn't indicate the most important thing - enemy outposts. And as I already wrote, they didn't need to fix it, it only helped. And if it got in the way, then you could put the order to “Follow” as first priority rather than "Fix Leaks".

_«- Fixed orders persisting even if the target no longer exists after a sub switch (e.g. a bot might attempt to operate a turret that doesn't exist, resulting in them doing nothing).»_
This is also pointless. The only thing I noticed was that on the new submarine the orders on the bots’ turrets “remained,” but they just had to be repeated and there were no problems. It didn't bother at all.

_«- Fixed "follow" order not always persisting between rounds.»_
What is this anyway? It was quite logical that, for example, in a friendly outpost or city, bot was given priority to “follow”, because we were, as it were, on “rest”. But now bots, once on the submarine (while at a friendly outpost), begins to do its job when it's not needed.

And apparently somewhere the developers made a mistake, which is why other orders broke (the same "Assault Enemy" - or they did it on purpose, I don't know).
